source/differentiation/ad/ad_drivers.cc

```diff
@@ -0,0 +1,2260 @@
+    std::vector<typename Types<ADNumberType>::tape_index>
+    TapedDrivers<
+      ADNumberType,
+      double,
+      typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                              NumberTypes::adolc_taped>::type>::
+      get_registered_tape_indices() const
+    {
+      // We've chosen to use unsigned shorts for the tape
+      // index type (a safety precaution) so we need to
+      // perform a conversion betwwen ADOL-C's native tape
+      // index type and that chosen by us.
+      std::vector<short> registered_tape_indices_s;
+      cachedTraceTags(registered_tape_indices_s);
+
+      std::vector<typename Types<ADNumberType>::tape_index>
+        registered_tape_indices(registered_tape_indices_s.size());
+      std::copy(registered_tape_indices_s.begin(),
+                registered_tape_indices_s.end(),
+                registered_tape_indices.begin());
+      return registered_tape_indices;
+    }
+
+
+    template <typename ADNumberType>
+    void
+    TapedDrivers<
+      ADNumberType,
+      double,
+      typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                              NumberTypes::adolc_taped>::type>::
+      activate_tape(const typename Types<ADNumberType>::tape_index tape_index)
+    {
+      active_tape = tape_index;
+    }
+
+
+    template <typename ADNumberType>
+    bool
+    TapedDrivers<
+      ADNumberType,
+      double,
+      typename std::enable_if<ADNumberTraits<ADNumberType>::type_code ==
+                              NumberTypes::adolc_taped>::type>::
+      requires_retaping(
+        const typename Types<ADNumberType>::tape_index tape_index) const
+    {
+      Assert(
+        is_registered_tape(tape_index) == true,
+        ExcMessage(
+          "Cannot ask for the status of a tape that has not yet been recorded and used."));
+
+      const auto it_status_tape = status.find(tape_index);
+
+      // This tape's status has not been found in the map. This could be
+      // because a non-existant tape_index has been used as an argument, or
+      // because the tape exists but has not been used (in a way that
+      // initiated a status update). For example, on can create a tape in one
+      // section of code and then query the status of all existing tapes in a
+      // completely different section of code that knows nothing about the
+      // first tape. There is no prerequisite that the first tape is ever
+      // used, and it can therefore not have a status. So, in this case
+      // there's not much we can do other than to report that the tape does
+      // not require retaping.
+      if (it_status_tape == status.end())
+        return false;
+
+      const auto status_tape = it_status_tape->second;
+
+      // See ADOL-C manual section 1.7 and comments in last paragraph of
+      // section 3.1
+      Assert(status_tape < 4 && status_tape >= -2,
+             ExcIndexRange(status_tape, -2, 4));
```

Comment:
I would rather we use `ExcInternalError()` or a custom message for this. That's because we use the exception you use here for user-provided indices, but here if you get something out of range, there is nothing a user could have done wrong...